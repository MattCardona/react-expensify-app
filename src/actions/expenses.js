import uuid from 'uuid';
import {database} from '../firebase/firebase.js';

// ADD_EXPENSE
const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {description = '', note = '', amount = 0, createdAt = 0} = expenseData;
    const expense = {description, note, amount, createdAt};

    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });
  };
};

// REMOVE_EXPENSE
const removeExpense = ({id} = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

const startRemoveExpense = ({id} = {}) => {
  console.log(`Hey this id: id from actions ${id.id}`);
  return (dispatch) => {
    return database.ref(`expenses/${id}`).remove().then(() => {
      console.log('Data was removed');
      dispatch(removeExpense({id}));
    }).catch((e) => {
      console.log(`Their was a error: ${e}`);
    });
  }
};

// EDIT_EXPENSE
const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});


//SET_EXPENSES
const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

const startSetExpenses = () => {
  return (dispatch) => {
    return database.ref('expenses').once('value').then((snapshot) => {
      const dbExpenses = [];
      snapshot.forEach((child) => {
        dbExpenses.push({
          id: child.key,
          ...child.val()
        });
      dispatch(setExpenses(dbExpenses));
      });
    });

  }
};

export {addExpense, startAddExpense, removeExpense, startRemoveExpense, editExpense, setExpenses, startSetExpenses};