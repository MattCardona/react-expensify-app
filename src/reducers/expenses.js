const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch(action.type) {
    case 'ADD_EXPENSE':
      return [...state,action.expense]
    case 'REMOVE_EXPENSE':
      return state.reduce((memo, expense) => {
        if(expense.id !== action.id){
          memo.push(expense);
          return memo;
        }
        return memo;
      }, []);
    case 'EDIT_EXPENSE':
      return state.reduce((memo, expense) => {
        if(expense.id === action.id){
          memo.push({...expense, ...action.updates});
          return memo;
        }
        memo.push(expense);
        return memo;
      }, []);
    default:
      return state;
  }
};

export {expensesReducer};