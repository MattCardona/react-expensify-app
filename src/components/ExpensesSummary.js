import React from 'react';
import {connect} from 'react-redux';
import {Link} from 'react-router-dom'
import {getVisibleExpenses} from '../selectors/expenses.js';
import {getExpenseTotal} from '../selectors/expenses-total.js';
import numeral from 'numeral';

const ExpensesSummary = (props) => {
  const expenseWord = props.expenseCount === 1 ? 'expense' : 'expenses';
  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">Viewing <span>{props.expenseCount}</span> {expenseWord},
        totalling <span>{props.expenseTotal}</span>
        </h1>
        <div className="page-header__actions">
          <Link className="button" to="/create">Add Expense</Link>
        </div>
      </div>
    </div>
)};

const mapStateToProps = (state) => {
  const expenses = getVisibleExpenses(state.expenses, state.filters);
  const total = numeral(getExpenseTotal(expenses)/ 100).format('$0,0.00');
  return {
        expenseCount: expenses.length,
        expenseTotal: total
      }
};

const ConnectedExpensesSummary = connect(mapStateToProps)(ExpensesSummary);

export {ExpensesSummary, ConnectedExpensesSummary};