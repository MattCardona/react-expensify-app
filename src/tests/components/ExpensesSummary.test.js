import React from 'react';
import {shallow} from 'enzyme';
import {ExpensesSummary} from '../../components/ExpensesSummary.js';
import expenses from '../fixtures/expenses.js';

test("should view 2 expenses totalling $4,695.00", () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={2} expenseTotal={'$4,695.00'} />);
  expect(wrapper).toMatchSnapshot();
});

test("should view 1 expense totalling $195.00", () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={1} expenseTotal={'$195.00'}/>);
  expect(wrapper).toMatchSnapshot();
});