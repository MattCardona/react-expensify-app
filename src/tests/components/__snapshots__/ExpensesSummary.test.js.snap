// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should view 1 expense totalling $195.00 1`] = `
<div>
  <h3>
    Viewing 
    1
     
    expense
    , totalling 
    $195.00
  </h3>
</div>
`;

exports[`should view 2 expenses totalling $4,695.00 1`] = `
<div>
  <h3>
    Viewing 
    2
     
    expenses
    , totalling 
    $4,695.00
  </h3>
</div>
`;
